pkg/extension/extension.go

```diff
@@ -0,0 +1,22 @@
+package extension
+
+import (
+	"context"
+
+	"helm.sh/helm/v3/pkg/chartutil"
+	"helm.sh/helm/v3/pkg/release"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+type Extension interface{}
+
+type PreReconciliationExtension interface {
+	PreReconcile(ctx context.Context, obj *unstructured.Unstructured, release release.Release, vals chartutil.Values) error
+}
+type PreDeletionExtension interface {
```

Comment:
There also need to be comments to explain the deletion workflow. In particular, it should be explained that if a CR exists but has a non-null deletion timestamp, `PreReconcile` above will _not_ be invoked, but `PreDelete` will (I'm btw still not convinced that the distinction between `PreReconcile` and `PreDelete` as potentially independent things makes sense - a correctly implemented extension not leaking resources that implements `PreReconcile` but doesn't need `PreDelete` seems not impossible, but very unlikely -, but I don't know what conversations you had with Joe Lanford to that end)